runtime/src/accounts_db.rs

```diff
@@ -2200,33 +2212,156 @@ impl AccountsDb {
+        let mut num_acceptable_failed_iterations = 0;
+        loop {
+            if num_acceptable_failed_iterations >= 10 {
+                // The latest version of the account existed in the index, but could not be
+                // fetched from storage. This means a race occurred between this function and clean
+                // accounts/purge_slots
+                datapoint_warn!(
+                    "accounts_db-do_load_warn",
+                    (
+                        "warn",
+                        format!("do_load() failed to get key: {} from storage, latest attempt was for slot: {}, storage_entry: {}",
+                        pubkey,
+                        slot,
+                        store_id),
+                        String
+                    ),
+                );
+                return None;
+            }
+            match account_accessor {
+                LoadedAccountAccessor::Stored(None) => {
+                    if is_transaction_load {
+                        panic!(
+                            "It should not happen that the storage entry doesn't exist if the entry in
+                            the accounts index is the latest version of this account. This is because clean
+                            should not be removing storage entries that have alive accounts"
+                        );
+                    }
+                    // RPC get_account() may have fetched an old root from the index that was
+                    // either:
+                    // 1) Cleaned up by clean_accounts(), so the accounts index has been updated
+                    // and the storage entries have been removed.
+                    // 2) Dropped by purge_slots() because the slot was on a minor fork, which
+                    // removes the slots' storage entries but doesn't purge from the accounts index
+                    // (account index cleanup is left to clean for stored slots).
+                    else {
+                        // Increment failsafe
+                        num_acceptable_failed_iterations += 1;
+                    }
+                }
+                LoadedAccountAccessor::Cached(None) => {
+                    // Cache was flushed in between checking the index and retrieving from the cache,
+                    // so retry. Don't increment `num_failed_iterations` here because it's impossible
+                    // for this to fail for transaction loads more than once. This is because for a slot
+                    // `X` that's being replayed, there is only one latest ancestor containing the latest
+                    // update for the account, and this ancestor can only be flushed once.
+                    num_acceptable_failed_iterations += 1;
+                    if is_transaction_load {
+                        assert!(num_acceptable_failed_iterations <= 1);
+                    }
+                }
+                _ => {
```

Comment:
nits: I think `LoadedAccountAccessor::Cached(Some(_) | LoadedAccountAccessor::Stored(Some(_))` won't hurt us for now so much, trading for future maintainability aided by rust's exhaustiveness check. :)